cli/src/commands/gerrit/upload.rs

```diff
@@ -0,0 +1,389 @@
+            .iter()
+            .filter(|trailer| trailer.key == "Change-Id")
+            .collect();
+
+        // There shouldn't be multiple change-ID fields. So just error out if
+        // there is.
+        if change_id_trailers.len() > 1 {
+            return Err(user_error(format!(
+                "multiple Change-Id footers in commit {}",
+                short_commit_hash(commit_id)
+            )));
+        }
+
+        // The user can choose to explicitly set their own change-ID to
+        // override the default change-ID based on the jj change-ID.
+        if let Some(trailer) = change_id_trailers.first() {
+            // Check the change-id format is correct.
+            if trailer.value.len() != 41 || !trailer.value.starts_with('I') {
+                // Intentionally leave the invalid change IDs as-is.
+                writeln!(
+                    ui.warning_default(),
+                    "warning: invalid Change-Id footer in commit {}",
+                    short_commit_hash(original_commit.id()),
+                )?;
+            }
+
+            // map the old commit to itself
+            old_to_new.insert(original_commit.id().clone(), original_commit.clone());
+            continue;
+        }
+
+        // Gerrit change id is 40 chars, jj change id is 32, so we need padding.
+        // To be consistent with `format_gerrit_change_id_trailer``, we pad with
+        // 6a6a6964 (hex of "jjid").
+        let gerrit_change_id = format!("I6a6a6964{}", original_commit.change_id().hex());
+
+        let new_description = format!(
+            "{}{}Change-Id: {}\n",
+            description.trim(),
+            if trailers.is_empty() { "\n\n" } else { "\n" },
+            gerrit_change_id
+        );
+
+        let new_parents = original_commit
+            .parents()
+            .map(|parent| {
+                let p = parent.unwrap();
+                if let Some(rewritten_parent) = old_to_new.get(p.id()) {
+                    rewritten_parent
+                } else {
+                    &p
+                }
+                .id()
+                .clone()
+            })
+            .collect();
+
+        // rewrite the set of parents to point to the commits that were
+        // previously rewritten in toposort order
+        //
+        // TODO FIXME (aseipp): this whole dance with toposorting, calculating
+        // new_parents, and then doing rewrite_commit is roughly equivalent to
+        // what we do in duplicate.rs as well. we should probably refactor this?
+        let new_commit = tx
+            .repo_mut()
+            .rewrite_commit(&original_commit)
+            .set_description(new_description)
+            .set_parents(new_parents)
+            // Set the timestamp back to the timestamp of the original commit.
+            // Otherwise, `jj gerrit upload @ && jj gerrit upload @` will upload
+            // two patchsets with the only difference being the timestamp.
```

Comment:
I think this may be a bug. See comment below about `tx.finish()`. Once we've added the `Change-Id` trailer, any subsequent upload is supposed to leave the commit unchanged, right? I.e. I don't think this rewrite should happen.